File: CBLClient/Apps/CBLTestServer-iOS/CBLTestServer-iOS/Server/CollectionRequestHandler.swift
//RequestHandler.swift
//CBLTestServer-iOS

import Foundation
import CouchbaseLiteSwift

public class CollectionRequestHandler {
    public func handleRequest(method: String, args: Args) throws -> Any? {
        switch method {
        case "collection_collection":
            let database: Database = args.get(name: "database")!
            let scopeName: String = args.get(name: "scopeName") ?? "_default"
            let collectionName: String = args.get(name: "collectionName")!
            return try database.collection(name: collectionName, scope: scopeName)
            
        case "collection_defaultCollection":
            let database: Database = (args.get(name:"database"))!
            return try database.defaultCollection()
            
        case "collection_getCollectionName":
            let collection: Collection = (args.get(name:"collection"))!
            return collection.name
            
        case "collection_createCollection":
            let scopeName: String = (args.get(name: "scopeName")) ?? "_default"
            let collectionName: String = (args.get(name: "collectionName"))!
            let database: Database = args.get(name: "database")!
            return try database.createCollection(name: collectionName, scope: scopeName)
            
        case "collection_collectionNames":
            let scope: String = (args.get(name:"scopeName")) ?? "_default"
            let database: Database = (args.get(name:"database"))!
            var names = [String]()
            let collectionNames = try database.collections(scope: scope)
            for collection in collectionNames{
                let collectionObject: Collection = collection
                names.append(collectionObject.name)
            }
            return names
            
        case "collection_documentCount":
            let collection: Collection = (args.get(name: "collection"))!
            return collection.count
            
        case "collection_collectionScope":
            let collection: Collection = args.get(name: "collection")!
            return collection.scope
            
        case "collection_collectionInstances":
            let scope: String = (args.get(name:"scopeName")) ?? "_default"
            let database: Database = (args.get(name:"database"))!
            return try database.collections(scope: scope)
        
        case "collection_deleteCollection":
            let scopeName: String = (args.get(name:"scopeName")) ?? "_default"
            let database: Database = (args.get(name:"database"))!
            let collectionName: String = (args.get(name:"collectionName"))!
            try database.deleteCollection(name: collectionName, scope: scopeName)
            
        case "collection_saveDocument":
            let document: MutableDocument = (args.get(name:"document"))!
            let collection: Collection = (args.get(name:"collection"))!
            try collection.save(document: document)

        case "collection_getDocument":
            let docId: String = (args.get(name: "docId"))!
            let collection : Collection = (args.get(name: "collection"))!
            return try collection.document(id: docId)
        
        case "collection_getDocuments":
            let collection: Collection = args.get(name:"collection")!
            let ids: [String] = args.get(name:"ids")!
            var documents = [String: [String: Any]]()

            for id in ids {
                let document: Document? = try collection.document(id: id)
                if document != nil{
                    var dict = document!.toDictionary()
                    for (key, value) in dict {
                        if let list = value as? Dictionary<String, Blob> {
                            var item = Dictionary<String, Any>()
                            for (k, v) in list {
                                item[k] = v.properties
                            }
                            dict[key] = item
                        }
                    }
                    documents[id] = dict
                }
            }

            return documents
        
        case "collection_updateDocument":
            let collection: Collection = args.get(name: "collection")!
            let data: Dictionary<String, Any> = args.get(name: "data")!
            let docId: String = args.get(name: "id")!
            let updated_doc = try collection.document(id: docId)!.toMutable()
            let new_data: Dictionary<String, Any> = setDataBlob(data)
            
            updated_doc.setData(new_data)
            try! collection.save(document: updated_doc)
            
        case "collection_deleteDocument":
            let document: Document = (args.get(name: "document"))!
            let collection: Collection = (args.get(name: "collection"))!
            try collection.delete(document: document)
            
        case "collection_getDocIds":
            let collection: Collection = args.get(name:"collection")!
            let limit: Int = args.get(name:"limit")!
            let offset: Int = args.get(name:"offset")!
            let query = QueryBuilder
                .select(SelectResult.expression(Meta.id))
                .from(DataSource.collection(collection))
                .limit(Expression.int(limit), offset:Expression.int(offset))

            var result: [String] = []
            do {
                for row in try query.execute() {
                    result.append(row.string(forKey: "id")!)
                }
            }

            return result

        case "collection_purgeDocument":
            let document: Document = (args.get(name: "document"))!
            let collection: Collection = (args.get(name: "collection"))!
            try collection.purge(document: document)
      
        case "collection_purgeDocumentID":
            let docID: String = (args.get(name: "docId"))!
            let collection: Collection = (args.get(name: "collection"))!
            try collection.purge(id: docID)

        case "collection_getDocumentExpiration":
            let docId: String = (args.get(name: "docId"))!
            let collection: Collection = (args.get(name: "collection"))!
            return try collection.getDocumentExpiration(id: docId)

        case "collection_setDocumentExpiration":
            let docId: String = (args.get(name: "docId"))!
            let expirationTime: Date = (args.get(name: "expirationTime"))!
            let collection: Collection = (args.get(name: "collection"))!
            try collection.setDocumentExpiration(id: docId, expiration: expirationTime)

        case "collection_getIndexNames":
            let collection: Collection = (args.get(name: "collection"))!
            return try collection.indexes()
            
        case "collection_deleteIndex":
            let collection: Collection = (args.get(name: "collection"))!
            let indexName: String = (args.get(name: "name"))!
            try collection.deleteIndex(forName: indexName)
            
        case "collection_createValueIndex":
            let collection: Collection = (args.get(name: "collection"))!
            let name: String = (args.get(name: "name"))!
            let expression: [String] = (args.get(name: "expression"))!
            let config = ValueIndexConfiguration(expression)
            try collection.createIndex(withName: name, config: config)          
            
        case "collection_saveDocuments":
            let database: Database = args.get(name:"database")!
            let collection: Collection = args.get(name: "collection")!
            let documents: Dictionary<String, Dictionary<String, Any>> = args.get(name: "documents")!
            try database.inBatch {
                for doc in documents {
                    let id = doc.key
                    var data: Dictionary<String, Any> = doc.value
                    if data["_id"] != nil {
                        data["id"] = id
                        data.removeValue(forKey: "_id")
                    }
                    let new_data: Dictionary<String, Any> = setDataBlob(data)
                    let document = MutableDocument(id: id, data: new_data)
                    try! collection.save(document: document)
                   }
            }
            
        default:
            throw RequestHandlerError.MethodNotFound(method)
        }
        return nil
    }
}
private extension CollectionRequestHandler {
    func setDataBlob(_ data: Dictionary<String, Any>) -> Dictionary<String, Any> {
        guard let attachment_items = data["_attachments"] as? Dictionary<String, Dictionary<String, Any>> else {
            return data
        }
        var existingBlobItems = [String: Any]()
        var updatedData = data;
        for (key, value) in attachment_items {
            
            if let d = value["data"] as? String  {
            
                let contentType = key.hasSuffix(".png") ? "image/jpeg" : "text/plain"
                let blob = Blob(contentType: contentType, data: d.data(using: .utf8)!)
                
                updatedData[key] = blob
                
            } else if let _ = value["digest"] {
                existingBlobItems[key] = value
            }
          }
        updatedData.removeValue(forKey: "_attachments")
        if !existingBlobItems.isEmpty {
            updatedData["_attachments"] = existingBlobItems
        }
        return updatedData
    }
}

File: CBLClient/Apps/JavaRequestHandlers/src/main/java/com/couchbase/mobiletestkit/javacommon/RequestHandler/CollectionRequestHandler.java
package com.couchbase.mobiletestkit.javacommon.RequestHandler;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.couchbase.lite.Blob;
import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.DataSource;
import com.couchbase.lite.Database;
import com.couchbase.lite.Expression;
import com.couchbase.lite.Index;
import com.couchbase.lite.IndexConfiguration;
import com.couchbase.lite.Limit;
import com.couchbase.lite.Meta;
import com.couchbase.lite.Query;
import com.couchbase.lite.QueryBuilder;
import com.couchbase.lite.Result;
import com.couchbase.lite.ResultSet;
import com.couchbase.lite.Scope;
import com.couchbase.lite.SelectResult;
import com.couchbase.lite.QueryIndex;
import com.couchbase.lite.ValueIndexConfiguration;
import com.couchbase.mobiletestkit.javacommon.Args;
import com.couchbase.lite.Collection;

import com.couchbase.lite.Document;
import com.couchbase.lite.MutableDocument;
import com.couchbase.mobiletestkit.javacommon.RequestHandlerDispatcher;
import com.couchbase.mobiletestkit.javacommon.util.Log;

public class CollectionRequestHandler {
    private static final String TAG = "DATABASE";
    /* _______________*/
    /* - Collection - */
    /* _______________*/
    public String getCollectionName(Args args) throws CouchbaseLiteException {
        Collection object = args.get("collection");
        return object.getName();
    }

    public List<Collection> collectionInstances(Args args) throws CouchbaseLiteException {
        String scopeName = (args.get("scopeName") != null) ? args.get("scopeName") : "_default";
        Database db = args.get("database");
        Set<Collection> setCollections = db.getCollections(scopeName);
        List<Collection> returnCollections = new ArrayList<Collection>();
        returnCollections.addAll(setCollections);
        return returnCollections;
    }

    public Collection createCollection(Args args) throws CouchbaseLiteException {
        Database db = args.get("database");
        String collectionName = args.get("collectionName");
        String scopeName = args.get("scopeName");
        if (scopeName == null) {
            return db.createCollection(collectionName);
        }
        else {
            return db.createCollection(collectionName, scopeName);
        }
    }

    public void deleteCollection(Args args) throws CouchbaseLiteException {
        Database db = args.get("database");
        String collectionName = args.get("collectionName");
        String scopeName = args.get("scopeName");
        if (scopeName == null) {
            db.deleteCollection(collectionName);
            return;
        }
        else {
            db.deleteCollection(collectionName, scopeName);
            return;
        }
    }

    public Collection defaultCollection(Args args) throws CouchbaseLiteException {
        Database db = args.get("database");
        return db.getDefaultCollection();

    }

    public static Set<Collection> collectionsInScope(Database db, String scopeName) throws CouchbaseLiteException {
        if (scopeName == null) {
            return db.getCollections();
        }
        else {
            return db.getCollections(scopeName);
        }
    }

    public ArrayList<String> collectionNames(Args args) throws CouchbaseLiteException {
        Database db = args.get("database");
        String scopeName = args.get("scopeName");
        ArrayList<String> collectionsNames = new ArrayList<String>();
        Set<Collection> collectionObjects = collectionsInScope(db, scopeName);
        for (Collection collectionObject: collectionObjects) {
            collectionsNames.add(collectionObject.getName());
        }
        return collectionsNames;
    }

    public List<String> getDocIds(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        int limit = args.get("limit");
        int offset = args.get("offset");
        Query query = QueryBuilder
                .select(SelectResult.expression(Meta.id))
                .from(DataSource.collection(collection))
                .limit(Expression.intValue(limit), Expression.intValue(offset));
        List<String> result = new ArrayList<>();
        ResultSet results = query.execute();
        for (Result row : results) {

            result.add(row.getString("id"));
        }
        return result;
    }

    public Map<String, Map<String, Object>> getDocuments(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        List<String> ids = args.get("ids");
        Map<String, Map<String, Object>> documents = new HashMap<>();
        for (String id : ids) {
            Document document = collection.getDocument(id);
            if (document != null) {
                Map<String, Object> doc = document.toMap();
                // looping through the document, replace the Blob with its properties
                for (Map.Entry<String, Object> entry : doc.entrySet()) {
                    if (entry.getValue() != null && entry.getValue() instanceof Map<?, ?>) {
                        if (((Map) entry.getValue()).size() == 0) {
                            continue;
                        }
                        boolean isBlob = false;
                        Map<?, ?> value = (Map<?, ?>) entry.getValue();
                        Map<String, Object> newVal = new HashMap<>();
                        for (Map.Entry<?, ?> item : value.entrySet()) {
                            if (item.getValue() != null && item.getValue() instanceof Blob) {
                                isBlob = true;
                                Blob b = (Blob) item.getValue();
                                newVal.put(item.getKey().toString(), b.getProperties());
                            }
                        }
                        if (isBlob) { 
                            doc.put(entry.getKey(), newVal); 
                        }
                    }
                }
                documents.put(id, doc);
            }
        }
        return documents;
    }

    public void updateDocument(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        String id = args.get("id");
        Map<String, Object> data = args.get("data");
        MutableDocument updatedDoc = collection.getDocument(id).toMutable();
        Map<String, Object> new_data = this.setDataBlob(data);
        updatedDoc.setData(new_data);
        collection.save(updatedDoc);
    }

    public long documentCount(Args args) throws CouchbaseLiteException {
        Collection collectionObject = args.get("collection");
        return collectionObject.getCount();
    }

    public Collection collection(Args args) throws CouchbaseLiteException {
        Database db = args.get("database");
        String collectionName = args.get("collectionName");
        String scopeName = args.get("scopeName");
        if (scopeName == null) {
            return db.getCollection(collectionName);
        }
        else {
            return db.getCollection(collectionName, scopeName);
        }
    }

    public void saveDocument(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        MutableDocument document = args.get("document");
        collection.save(document);
        return;
    }

    public Scope collectionScope(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        return collection.getScope();
    }

    public Document getDocument(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        String docId = args.get("docId");
        return collection.getDocument(docId);
    }

    public void deleteDocument(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        Document doc = args.get("document");
        collection.delete(doc);
        return;
    }

    public void purgeDocument(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        Document doc = args.get("document");
        collection.purge(doc);
        return;
    }

    public void purgeDocumentID(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        String docID = args.get("docId");
        collection.purge(docID);
        return;
    }

    public Date getDocumentExpiration(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        String docId = args.get("docId");
        return collection.getDocumentExpiration(docId);
    }

    public void setDocumentExpiration(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        String docId = args.get("docId");
        Date expiration = args.get("expiration");
        collection.setDocumentExpiration(docId, expiration);
        return;
    }

    public void createValueIndex(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        String name = args.get("name");
        String expression = args.get("expression");
        IndexConfiguration expressionConfig = new ValueIndexConfiguration(expression);
        collection.createIndex(name, expressionConfig);
        return;
    }

    public QueryIndex getIndex(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        String name = args.get("indexName");
        return collection.getIndex(name);
    }

    public void deleteIndex(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        String name = args.get("name");
        collection.deleteIndex(name);
        return;
    }

    public Set<String> getIndexNames(Args args) throws CouchbaseLiteException {
        Collection collection = args.get("collection");
        return collection.getIndexes();
    }

    public void saveDocuments(Args args) throws CouchbaseLiteException {
        final Collection collection = args.get("collection");
        final Database db = args.get("database");
        final Map<String, Map<String, Object>> documents = args.get("documents");
        db.inBatch(() -> {
            for (Map.Entry<String, Map<String, Object>> entry : documents.entrySet()) {
                String id = entry.getKey();
                Map<String, Object> data = entry.getValue();
                Map<String, Object> new_data = this.setDataBlob(data);
                MutableDocument document = new MutableDocument(id, new_data);
                try {
                    collection.save(document);
                }
                catch (CouchbaseLiteException e) {
                    Log.e(TAG, "DB Save failed", e);
                }
            }
        });
    }

    private Map<String, Object> setDataBlob(Map<String, Object> data) {
        if (!data.containsKey("_attachments")) {
            return data;
        }

        Map<String, Object> attachment_items = (Map<String, Object>) data.get("_attachments");
        Map<String, Object> existingBlobItems = new HashMap<>();

        for (Map.Entry<String, Object> attItem : attachment_items.entrySet()) {
            String attItemKey = attItem.getKey();
            HashMap<String, String> attItemValue = (HashMap<String, String>) attItem.getValue();
            if (attItemValue.get("data") != null) {
                String contentType = attItemKey.endsWith(".png") ? "image/jpeg": "text/plain";
                Blob blob = new Blob(contentType,
                        RequestHandlerDispatcher.context.decodeBase64(attItemValue.get("data")));
                data.put(attItemKey, blob);

            }
            else if (attItemValue.containsKey("digest")) {
                existingBlobItems.put(attItemKey, attItemValue);
            }
        }
        data.remove("_attachments");
        // deal with partial blob situation,
        // remove all elements then add back blob type only items to _attachments key
        if (existingBlobItems.size() > 0 && existingBlobItems.size() < attachment_items.size()) {
            data.remove("_attachments");
            data.put("_attachments", existingBlobItems);
        }

        return data;
    }
}